Allow instantiation from class methods

dhpollack opened the pull request:
* instantiate an object from a class method of that object
* add tests for this
* simplify get_static_method function

Signed-off-by: David Pollack <d.pollack@solvemate.com>

## Motivation

I wanted to use hydra with the huggingface transfromers library and they extensively use a class method named `from_pretrained`.  

Currently one could do this by using the `get_static_method` and then manually converting the `params` key and passing them to the gotten method.  However that's not very straight forward or intuitive and this doesn't add much logic to the utility function.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes - passed tests in test_utils.py and flake8

## Test Plan

I added a test that tests the basic case.

omry commented on tests/test_utils.py:
```diff
@@ -118,6 +124,16 @@ def test_get_static_method(path: str, return_value: Any) -> None:
+        (
+            {
+                "cls": "tests.test_utils.Baz",
+                "method": "class_method",
```
The syntax I would like to support is:
```yaml
foo:
   cls: tests.test_utils.Baz.class_method
```
Is this implementable?
You may need to differentiate which of the two modes we are in automatically.

dhpollack replied:
Yea, I wanted that too, but it didn't seem easy if you use the `instantiate` method because you would have to always check the last and second to last items to see if the last item in the string was a method or a class of the method of the second to last item.  I thought it would create extra logic that made things more confusing.  My other idea was to create a `instantiate_from_class_method` function (or whatever you'd want it to be called) which assumes that form.  Should work either way, it just depends where you want to put the logic.  

I'll take a look at it and see if I can put it all into the `cls` key neatly.

dhpollack replied:
Yea, I wanted that too, but it didn't seem easy if you use the `instantiate` method because you would have to always check the last and second to last items to see if the last item in the string was a method or a class of the method of the second to last item.  I thought it would create extra logic that made things more confusing.  My other idea was to create a `instantiate_from_class_method` function (or whatever you'd want it to be called) which assumes that form.  Should work either way, it just depends where you want to put the logic.  

I'll take a look at it and see if I can put it all into the `cls` key neatly.

omry replied:
Yeah, the implementation would be more complicated but I think it's okay.
Let's see what it looks like.

By the way, I just realized that PluginConf is now a part of the public API. I originally thought of it as an internal Hydra detail. I will have to promote it to be a part of the API. (probably also rename it).

dhpollack replied:
@omry alright, it now works.  I used the the `locate` from from `pydoc` which basically does all the work.  I also simplified the other parts of the code that were duplicating functionality of this such as your static method function.

omry commented on hydra/utils.py:
```diff
@@ -18,32 +18,24 @@ def get_method(path: str) -> type:
-            )
+        from pydoc import locate
+
+        klass = locate(path)
```
This can't be implemented with importlib?

dhpollack replied:
Under the hood it is implemented with `__import__` but with a bunch of logic to get the class methods and to recursive import the correct module.  I believe `importlib` is just a friendly wrapper around `__import__`.  You can see the code for [locate](https://github.com/python/cpython/blob/master/Lib/pydoc.py#L1620) and [safeimport (the underlying importer)](https://github.com/python/cpython/blob/master/Lib/pydoc.py#L318).  One could implement this with `importlib`. but it's not going to be as good as this implementation.

dhpollack replied:
I ended up re-implementing and simplifying the pydoc function because there was some strange behavior if a submodule on the chain doesn't load.

omry commented on hydra/utils.py:
```diff
@@ -116,9 +108,7 @@ def _instantiate_class(
-    ), "Input config params are expected to be a mapping, found {}".format(
-        type(config.params)
-    )
+    ), f"Input config params are expected to be a mapping, found {type(config.params)}"
```
```suggestion
    ), f"Input config params are expected to be a mapping, found {type(config.params).__name__}"
```

omry commented on tests/test_utils.py:
```diff
@@ -59,6 +59,12 @@ def __eq__(self, other: Any) -> Any:
 
 
+class Baz(Foo):
+    @classmethod
```
can you test it on `a @staticmethod ` as well?

dhpollack replied:
done

omry commented on website/docs/patterns/objects.md:
```diff
@@ -49,7 +53,8 @@ conf/
     ├── mysql.yaml
-    └── postgresql.yaml
+    ├── postgresql.yaml
+    └── postgresql_default.yaml
```
I think we don't need a complete example.
It's enough to say in the main doc that this supports:
 - classes
 - class methods
 - static methods
 - (and if you test) static class methods.

dhpollack replied:
Ok, where in the main doc is this?  I can add that to this PR.

omry replied:
Does not exist yet, see my comment. we can add a new section to objects.md for now.

omry commented:
As for the failing CI:
see the developer guide for info on how to run the tests and the lint locally.

omry commented on hydra/utils.py:
```diff
@@ -12,38 +13,82 @@
+        module = import_module(path)
+    except ImportError as e:
+        if e.name == path:
+            return None
```
what case is this handling?

dhpollack replied:
this was a relic from the pydoc method which checks the method from front to back, which is required if one uses `__import__`.  I removed it now.

omry commented on hydra/utils.py:
```diff
@@ -12,38 +13,82 @@
+            raise e
+    except Exception as e:
+        log.error(f"Non-ImportError while importing module {path}: {e}")
+        raise ValueError(f"Non-ImportError while importing module {path}: {e}")
```
the log.error is probably going to cause the message to be printed twice.

dhpollack replied:
You're right. I took it from the original file but I've noticed you've only used the log for BaseException errors that don't have a message attached.

omry commented on hydra/utils.py:
```diff
@@ -12,38 +13,82 @@
-            raise ImportError(
-                "Class {} is not in module {}".format(class_name, module_path)
-            )
+        klass = cast(type, _locate(path))
```
try : 

```python
t = _locate(path)
assert isinstance(t, type)
return t
```

dhpollack replied:
I tried this but it didn't work.  The return types of static methods and class methods were different than class objects.  I tried explicitly typing `t` with `t: type`, but assert statement still failed.  I also tried changing the assert statement to `Callable` instead of `type`, but then I recieved a failure from the mypy check.  In this case, all the tests passed.

I also tried changing all of the return types to `ModuleType` and adjust the function signature of `_instantiate_class` accordingly, but this caused more mypy errors and some of the tests started failing as well.

omry replied:
_locate signature suggests that the return type is ModuleType.
Since this function is now used for more than returning classes, maybe we need to refactor things such that it only get called for actual classes (by the user) and the internal implementation operates on ModuleType.

Does it make sense?
btw, can you join the chat?

dhpollack replied:
I am in Europe so I can join a chat today (for me, probably tomorrow for you) if you're free.  I just registered for the chat, so I'll be in there.

omry commented on hydra/utils.py:
```diff
@@ -12,38 +13,82 @@
-    except ValueError as e:
-        log.error("Error initializing class " + path)
+    except Exception as e:
+        log.error(f"Error initializing class {path}")
```
try:
```python
log.error(f"Error initializing class {path} : {e}")
```

omry commented on website/docs/patterns/objects.md:
```diff
@@ -33,6 +33,10 @@ class PostgreSQLConnection(DBConnection):
+    
+    @classmethod
+    def default_init_method(self, database):
+        return self("localhost", "admin", "password", database)
```
I think this can be removed now.

omry replied:
I propose that you add a new section to this page describing `hydra.utils.instantiate()`.
It's probably should come near the end, after we introduce it for the first time.

omry commented:
ping

dhpollack commented:
The test failues look like they are due to the upgrade in omegaconf.  Do you prefer a `git rebase` or a `git merge` on the master?  I tested both on branches locally.  I would assume rebasing, but I figured I'd ask.

omry commented:
I prefer rebase.

omry commented on website/docs/patterns/objects.md:
```diff
@@ -95,3 +95,7 @@ Change the instantiated object class and override values from the command line:
+
+In addition to instantiating objects from classes, the `instantiate` method can also
+be used for class and static methods.  Simply put the path to the method in the `cls`
+field and use `params` if needed and instatiate your object exactly like a normal class.
```
Please describe each of the use cases in more details.
Something along those lines (suggestion, feel free to diverge):

#### class names
An object of that class will be instantiated and returned. passing the config arguments to the `__init__` method
Example class:
```python
class Foo:
  def __init__(x: int, y: int) -> None:
    self.x = x
    self.y = y
```
Example usage:
config:
```yaml
object:
  cls: Foo
  params:
    x: 10
    y: 20
```
```python
result = hydra.utils.instantiate(cfg.object)
```

#### static_method
TODO: similar small example

#### class method
TOOD: similar small example

omry commented on tests/test_utils.py:
```diff
@@ -59,6 +59,16 @@ def __eq__(self, other: Any) -> Any:
+        return self(y + 1)
+
+    @staticmethod
+    def static_method() -> Any:
```
pass some parameters to the static method as well.

omry commented on hydra/utils.py:
```diff
@@ -133,5 +162,25 @@ def _instantiate_class(
+instantiate = call
+get_static_method = get_fn_or_method
+get_method = get_fn_or_method
+_get_class_name = _get_cls_name
```
There is only one usage of `_get_class_name()` in Hydra outside of this class. just fix it. no need to introduce an alias.

dhpollack replied:
:+1:

omry commented on website/docs/patterns/objects.md:
````diff
@@ -95,3 +95,62 @@ Change the instantiated object class and override values from the command line:
+  mystaticmethod = hydra.utils.call(cfg.mystaticmethod)
+  myfunction = hydra.utils.call(cfg.myfunction)
+```
+Note, the old `instantiate` function is now an alias of the `call` function used above.
````
```suggestion
Note: `hydra.utils.instantiate()` is an alias for `hydra.utils.call()`. They are in fact the same function.
```

dhpollack replied:
done

dhpollack commented:
Alright, I think that I solved the mypy issues by using `_instantiate_class` for classes and `_call_callable` for functions and methods.  I also created a bunch of aliases for the old naming scheme and renamed functions as well.  I also added a few extra tests for builtins and nested methods.  

As for the documentation, let me know what you think.  I tried to keep it as compact as possible, but may be doing too many things with the same class.

P.S.  the win37 failure looks unrelated to this PR.  Looks like the conda servers glitched.

omry commented on website/docs/patterns/objects.md:
```diff
@@ -95,3 +95,63 @@ Change the instantiated object class and override values from the command line:
+  myobject = hydra.utils.call(cfg.myobject)
+  myclassmethod = hydra.utils.call(cfg.myclassmethod)
+  mystaticmethod = hydra.utils.call(cfg.mystaticmethod)
+  myfunction = hydra.utils.call(cfg.myfunction)
```
```suggestion
  foo1: Foo = hydra.utils.call(cfg.myobject)
  foo2: Foo = hydra.utils.call(cfg.myclassmethod)
  ret1: int = hydra.utils.call(cfg.mystaticmethod)
  ret2: int = hydra.utils.call(cfg.myfunction)
```

omry commented:
As for the docs, they look good here but I would like to see them in the website.
Did you start the website locally to test that they look okay?

dhpollack commented:
I didn't but I'll check it tomorrow

omry commented:
Awesome. edited the doc a bit and added the issue #505 for API cleanup before 1.0.

omry commented:
Ohh, forgot.
Can you add a news fragment file?
see the developer guide for details.

omry commented:
Awesome, thanks David!
